src/lib.rs

```diff
@@ -172,27 +172,64 @@ fn encode_signature_url_safe(bytes: &[u8]) -> String {
 /// ```
 /// use cloudfront_url_signer;
 ///
-/// fn main() {
-///    let resource = "https://example.cloudfront.net/flowerpot.png";
-///    let expiry = 1579532331;
-///    let certificate_location = "examples/key.pem";
-///    let key_pair_id = "APKAIEXAMPLE";
-///
-///    let signature = cloudfront_url_signer::create_canned_policy_signature(resource, expiry, certificate_location).unwrap();
+///let resource = "https://example.cloudfront.net/flowerpot.png";
+///let expiry = 1579532331;
+///let certificate_location = "examples/key.pem";
+///let key_pair_id = "APKAIEXAMPLE";
+///let signature = cloudfront_url_signer::create_canned_policy_signature(resource, expiry, certificate_location).unwrap();
 ///
-///    println!("Signed URL is {}", format!("{}?Expires={}&Signature={}&Key-Pair-Id={}", resource, expiry, signature, key_pair_id));
-///}
+///println!("Signed URL is {}", format!("{}?Expires={}&Signature={}&Key-Pair-Id={}", resource, expiry, signature, key_pair_id));
 /// ```
 ///
 pub fn create_canned_policy_signature(
     resource: &str,
     expiry: u64,
     private_key_location: &str,
 ) -> Result<String, Error> {
-    read_rsa_private_key(private_key_location).and_then(|key| {
-        parse_rsa_private_key(&key).and_then(|private_key| {
-            sign_canned_policy(&generate_canned_policy(resource, expiry), &private_key)
-                .and_then(|signed_policy| Ok(encode_signature_url_safe(&signed_policy)))
+    let key = read_file_to_private_key(private_key_location)?;
+    let signed_policy = sign_canned_policy(&generate_canned_policy(resource, expiry), &key)?;
+
+    Ok(encode_signature_url_safe(&signed_policy))
+}
+
+pub fn read_file_to_private_key(private_key_location: &str) -> Result<PKey<Private>, Error> {
+    let key = read_rsa_private_key(private_key_location)?;
+
+    parse_rsa_private_key(&key)
+}
+
+/// Struct to create a URL from CloudFront with a cached private key
+/// Unliked method `create_canned_policy_signature`, calling `create_canned_policy_signature_url`
+/// in an instance of this struct will not read the private key file every time it is invoked.
+pub struct CloudFrontCannedPolicySigner {
```

Comment:
Omits reading the private key from the file system everytime someone wants to generate a url